File: wallendar-app/app/src/main/java/com/td/wallendar/addmembers/ui/AddMembersActivity.java
package com.td.wallendar.addmembers.ui;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textfield.TextInputLayout;
import com.td.wallendar.AbstractActivity;
import com.td.wallendar.R;
import com.td.wallendar.addmembers.MembersAdapter;
import com.td.wallendar.di.DependenciesContainer;
import com.td.wallendar.di.DependenciesContainerLocator;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class AddMembersActivity extends AbstractActivity implements AddMembersView {
    private AddMembersPresenter addMembersPresenter;

    private TextInputLayout memberInputLayout;
    private Button checkToAddMember;
    private RecyclerView recyclerView;

    private MembersAdapter membersAdapter;
    private final List<String> members = new ArrayList<>();

    private final long NO_GROUP_ID = -1;
    private final String GROUP_ID = "GROUP_ID";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_add_members);

        setupAdapter();
        setupActionBar();

        setupInput();
        setupAddMember();
        setupSubmitMembers();
        setupRecyclerView();

        createPresenter();
    }

    private void setupSubmitMembers() {
        final long groupId = getIntent().getLongExtra("GROUP_ID", -1);
        if (groupId == NO_GROUP_ID) {
            // TODO error
        } else {
            findViewById(R.id.submit_members).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    addMembersPresenter.submitMembers(groupId, members);
                }
            });
        }
    }

    private void setupAdapter() {
        membersAdapter = new MembersAdapter(members);
    }

    private void setupRecyclerView() {
        recyclerView = findViewById(R.id.current_members_added);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext(), LinearLayoutManager.VERTICAL, false));
        recyclerView.setAdapter(membersAdapter);
    }

    private void setupAddMember() {
        checkToAddMember = findViewById(R.id.add_member_button);
        checkToAddMember.setOnClickListener(view -> {
            final String currentMember = Objects.requireNonNull(memberInputLayout.getEditText()).getText().toString();
            memberInputLayout.getEditText().getText().clear();
            membersAdapter.addToDataset(currentMember);
        });
    }

    private void setupInput() {
        memberInputLayout = findViewById(R.id.member_layout);
    }

    private void createPresenter() {
        addMembersPresenter = (AddMembersPresenter) getLastNonConfigurationInstance();

        if (addMembersPresenter == null) {
            final DependenciesContainer dependenciesContainer = DependenciesContainerLocator.locateComponent(this);
            final GroupsRepository groupsRepository = dependenciesContainer.getGroupsRepository();
            final SchedulerProvider schedulerProvider = dependenciesContainer.getSchedulerProvider();
            addMembersPresenter = new AddMembersPresenter(this, groupsRepository, schedulerProvider);
        }
    }


    private void setupActionBar() {
        ActionBar actionBar = getSupportActionBar();

        assert actionBar != null;

        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_close_white);
        actionBar.setTitle(R.string.add_members);
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    // TODO we are not using the groupId param right now, may be useful in some borders cases
    @Override
    public void onMembersAddedSuccessfully(long groupId) {
        final Intent resultIntent = new Intent();
        setResult(RESULT_OK, resultIntent);
        finish();
    }

    @Override
    public void onMembersAddedWithError() {
        Toast.makeText(getApplicationContext(), "There was an error adding members", Toast.LENGTH_LONG).show();
        onBackPressed();
    }

    @Override
    public void onStop() {
        super.onStop();
        addMembersPresenter.onViewDetached();
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/addmembers/ui/AddMembersActivity.kt
package com.td.wallendar.addmembers.ui

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.Toast
import androidx.annotation.VisibleForTesting
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textfield.TextInputLayout
import com.td.wallendar.AbstractActivity
import com.td.wallendar.R
import com.td.wallendar.addmembers.MembersAdapter
import com.td.wallendar.di.DependenciesContainerLocator
import java.util.*

class AddMembersActivity : AbstractActivity(), AddMembersView {
    private var addMembersPresenter: AddMembersPresenter? = null
    private var memberInputLayout: TextInputLayout? = null
    private var checkToAddMember: Button? = null
    private var recyclerView: RecyclerView? = null
    private var membersAdapter: MembersAdapter? = null
    private val members: MutableList<String> = ArrayList()
    private val NO_GROUP_ID: Long = -1
    private val GROUP_ID: String = "GROUP_ID"
    private val IS_EVENT: String = "IS_EVENT"
    private var isEvent: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_members)
        setupAdapter()
        setupActionBar()
        setupInput()
        setupAddMember()
        setupSubmitMembers()
        setupRecyclerView()
        createPresenter()
    }

    private fun setupSubmitMembers() {
        val groupId = intent.getLongExtra(GROUP_ID, -1)
        if (groupId == NO_GROUP_ID) {
            // TODO error
        } else {
            findViewById<View?>(R.id.submit_members).setOnClickListener { addMembersPresenter?.submitMembers(groupId, members) }
        }
    }

    private fun setupAdapter() {
        membersAdapter = MembersAdapter(members)
    }

    private fun setupRecyclerView() {
        recyclerView = findViewById(R.id.current_members_added)
        recyclerView?.layoutManager = LinearLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)
        recyclerView?.adapter = membersAdapter
    }

    private fun setupAddMember() {
        checkToAddMember = findViewById(R.id.add_member_button)
        checkToAddMember?.setOnClickListener {
            val currentMember = Objects.requireNonNull(memberInputLayout?.editText)?.text.toString()
            if (currentMember.isNotEmpty()) {
                memberInputLayout?.editText?.text?.clear()
                membersAdapter?.addToDataset(currentMember)
            } else {
                Toast.makeText(applicationContext, "You must write an email.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun setupInput() {
        memberInputLayout = findViewById(R.id.member_layout)
    }

    private fun createPresenter() {
        addMembersPresenter = lastNonConfigurationInstance as AddMembersPresenter?
        if (addMembersPresenter == null) {
            val dependenciesContainer = DependenciesContainerLocator.locateComponent(this)
            val groupsRepository = dependenciesContainer.getGroupsRepository()
            val eventsRepository = dependenciesContainer.getEventsRepository()
            val schedulerProvider = dependenciesContainer.getSchedulerProvider()
            isEvent = intent.extras?.getBoolean(IS_EVENT)!!
            addMembersPresenter = AddMembersPresenter(this, groupsRepository, eventsRepository, schedulerProvider, isEvent)
        }
    }

    @VisibleForTesting
    fun setPresenter(addMembersPresenter: AddMembersPresenter?) {
        this.addMembersPresenter = addMembersPresenter
    }

    @VisibleForTesting
    fun getPresenter(): AddMembersPresenter? {
        return addMembersPresenter
    }

    override fun onRetainCustomNonConfigurationInstance(): Any? {
        return addMembersPresenter
    }

    private fun setupActionBar() {
        val actionBar = supportActionBar!!
        actionBar.setDisplayHomeAsUpEnabled(true)
        actionBar.setHomeAsUpIndicator(R.drawable.ic_close_white)
        actionBar.setTitle(R.string.add_members)
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }

    // TODO we are not using the groupId param right now, may be useful in some borders cases
    override fun onMembersAddedSuccessfully() {
        val resultIntent = Intent()
        setResult(RESULT_OK, resultIntent)
        finish()
    }

    override fun onMembersAddedWithError() {
        Toast.makeText(applicationContext, "There was an error adding members", Toast.LENGTH_SHORT).show()
        onBackPressed()
    }

    public override fun onStop() {
        super.onStop()
        addMembersPresenter?.onViewDetached()
    }
}